### PR TITLE
Lightning Fast

### PR DESCRIPTION
**Pull request by SergeyTsimfer**
This PR proposes to add:

- converter from `Pipeline` to `torch.DataLoader` to make it fully compatible with `torch`  features

- converted from `TorchModel` to `pytorch_lightning.LightningModule` so it can enjoy these features as well

### CODE REVIEW

File: batchflow/models/torch/base.py [hunk 1]
```diff
@@ -1573,3 +1583,35 @@ def close(self):
...
         self.close()
+
+
+
+class LightningModel(LightningModule):
+    """ Convert `TorchModel` from `BatchFlow` to `PyLightning`.
+    Falcon WA, et al "`PyTorch Lightning <https://github.com/PyTorchLightning/pytorch-lightning>`_"
+    """
+    def __init__(self, batchflow_model):
+        if LightningModule == object:
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1594]: Why not `None`?

File: batchflow/pipeline.py [hunk 2]
```diff
@@ -1637,3 +1642,39 @@ def run_now(self, *args, **kwargs):
...
+            Desired number of items in the batch.
+        kwargs : dict
+            Additional parameters of `DataLoader` initialization.
+        """
+        pipeline_dataset = PipelineDataset(self, *components, batch_size=batch_size)
+        return DataLoader(pipeline_dataset, batch_size=None, pin_memory=True, **kwargs)
+
+
+
+class PipelineDataset(IterableDataset):
```
Comment (roman-kh) [inline batchflow/pipeline.py:1664]: Let's move it to a separate module with `torch` in its name

File: batchflow/pipeline.py [hunk 3]
```diff
@@ -1637,3 +1642,39 @@ def run_now(self, *args, **kwargs):
...
+        ----------
+        components : sequence of str
+            Components to get from generated batches.
+        batch_size : int
+            Desired number of items in the batch.
+        kwargs : dict
+            Additional parameters of `DataLoader` initialization.
+        """
+        pipeline_dataset = PipelineDataset(self, *components, batch_size=batch_size)
+        return DataLoader(pipeline_dataset, batch_size=None, pin_memory=True, **kwargs)
```
Comment (roman-kh) [inline batchflow/pipeline.py:1660]: Make this method one-liner

File: batchflow/models/torch/base.py [hunk 4]
```diff
@@ -1573,3 +1583,35 @@ def close(self):
...
+
+    def forward(self, x):
+        return self.batchflow_model.model(x)
+
+    def configure_optimizers(self):
+        return self.batchflow_model.train_steps['']['optimizer']
+
+    def training_step(self, batch, batch_idx):
+        """ Define custom training iteration on given data. """
+        images, targets = batch
```
Comment (roman-kh) [inline batchflow/models/torch/base.py:1609]:
```suggestion
        inputs, targets = batch
```

**Review by roman-kh — CHANGES_REQUESTED** [review]

**Comment by nikita-klsh** [comment]
We need to unify optional packages imports across modules and files.

Here you use the block
```
try:
    from pytorch_lightning import LightningModule
except ImportError:
    LightningModule = object
```
And explicitly asking the user to install package in case of absence.

In `monitor.py` there is the block
```
try:
    import nvidia_smi
except ImportError:
    pass
```
And we dont ask to install anything. The `ImportError` implicitly raised in case of `nvidia-smi` absence

File: batchflow/pipeline.py [hunk 5]
```diff
@@ -1637,3 +1642,39 @@ def run_now(self, *args, **kwargs):
...
+
+
+
+class PipelineDataset(IterableDataset):
+    """ Wrap pipeline into `PyTorch` interface. """
+    def __init__(self, pipeline, *components, batch_size=64):
+        if IterableDataset == object:
```
Comment (anton-br) [inline batchflow/pipeline.py:1667]: what about import error for DataLoader?

File: batchflow/models/torch/lightning.py [hunk 6]
```diff
@@ -0,0 +1,86 @@
+""" Utilities to seamlessly convert `BatchFlow` model and data generation pipeline to
+`Pytorch Lightning` compatible modules to leverage all of its features.
+
+>>> lightning_model = batchflow_model.to_lightning()
+>>> dataloader = ppl.to_dataloader('images', 'labels', batch_size=64)
```
Comment (roman-kh) [inline batchflow/models/torch/lightning.py:5]: dataloader = pipeline.to_dataloader(batch_size=64, components=['images', 'labels'])

File: batchflow/models/torch/lightning.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
+""" Utilities to seamlessly convert `BatchFlow` model and data generation pipeline to
+`Pytorch Lightning` compatible modules to leverage all of its features.
+
+>>> lightning_model = batchflow_model.to_lightning()
```
Comment (roman-kh) [inline batchflow/models/torch/lightning.py:4]: There is still no way to convert a batchflow model to a custom lightning module class.

**Review by roman-kh — CHANGES_REQUESTED** [review]

File: batchflow/models/torch/lightning.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            batch = self.pipeline.next_batch(self.batch_size)
+            yield [getattr(batch, comp).astype(np.float32) for comp in self.components]
+
+    def __iter__(self):
+        return self.get_data()
+
+
+class PipelineDataLoader(DataLoader):
+    """ Create dataloader from `Pipeline` or `PipelineDataset`. """
+    def __init__(self, dataset, components=None, batch_size=None, pin_memory=True, **kwargs):
```
Comment (anton-br) [inline batchflow/models/torch/lightning.py:46]: add docs about init's arguments
